libtiny_tui/src/input_area/input_line.rs

```diff
@@ -0,0 +1,503 @@
+                **next_split.unwrap_or(&&self.buffer.len()) as i32 - char_idx;
+
+            // debug!("rendered_height: slots_in_line: {}, chars_until_next_split: {}",
+            //        slots_in_line, chars_until_next_split);
+            if chars_until_next_split > slots_in_line {
+                // debug!("splitting at {}", char_idx);
+                line_count += 1;
+                line_start = char_idx + 1;
+                line_width = width;
+            }
+        }
+        if self.word_splits.len() == 0 {
+            // no splits...must be one giant spaceless line
+            // get line with nickpadding
+            let first_line = self.len() as i32 - line_width;
+            if first_line >= 0 {
+                // get lines as wide as width
+                line_count += 1;
+                line_count += first_line / width;
+            }
+        }
+        line_count as usize
+    }
+}
+
+fn draw_line_scroll(
+    tb: &mut Termbox,
+    colors: &Colors,
+    line: &[char],
+    pos_x: i32,
+    pos_y: i32,
+    width: i32,
+    cursor: i32,
+    scroll: i32,
+) {
+    let slice: &[char] = &line[scroll as usize..min(line.len(), (scroll + width + pos_x) as usize)];
+    termbox::print_chars(tb, pos_x, pos_y, colors.user_msg, slice.iter().cloned());
+    // On my terminal the cursor is only shown when there's a character
+    // under it.
+    if cursor as usize >= line.len() {
+        tb.change_cell(
+            pos_x + cursor - scroll,
+            pos_y,
+            ' ',
+            colors.cursor.fg,
+            colors.cursor.bg,
+        );
+    }
+
+    tb.set_cursor(Some(((pos_x + cursor - scroll) as u16, pos_y as u16)));
+}
+
+pub(crate) struct CompletionRange {
+    start_idx: usize,
+    end_idx: usize,
+}
+
+fn draw_line_wrapped(
+    tb: &mut Termbox,
+    colors: &Colors,
+    pos_x: i32,
+    pos_y: i32,
+    width: i32,
+    cursor: i32,
+    line: &InputLine,
+    completion_range: &Option<CompletionRange>,
+) {
+    let mut col = pos_x;
+    let mut line_num = 0;
+
+    let mut cursor_xychar = (0, 0, ' ');
+    let mut check_cursor = |char_idx: usize, cursor: i32, pos_x: i32, pos_y: i32, c: char| {
+        if char_idx == cursor as usize {
+            cursor_xychar = (pos_x, pos_y, c);
+        }
+    };
+    let mut char_idx: usize = 0;
+    let buffer_length = line.buffer.len();
+    let mut split_iter = line.word_splits.iter().skip(1);
+    for c in &line.buffer {
+        let mut style = colors.user_msg;
+        // for autocompletion highlighting
+        if let Some(completion_range) = completion_range {
+            if char_idx >= completion_range.start_idx && char_idx < completion_range.end_idx {
+                style = colors.completion;
+            }
+        }
+        if c.is_whitespace() {
+            // We may want to move to the next line
+            let next_split = split_iter.next().unwrap_or_else(|| &buffer_length);
```

Comment:
`unwrap_or(&buffer_length)` would be better here as `buffer_length` is just a variable here, we don't do any computation in the closure.

Reply:
Ok, will change. I thought that unwrap_or_else was better since it lazily evaluates. Before, I had a debug message in there and it always printed, not only when the unwrap was `None`